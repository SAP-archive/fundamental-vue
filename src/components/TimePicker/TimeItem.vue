<template>
  <div class="fd-time">
    <Time
      :type="showMeridian ? 'hour12' : 'hour24'"
      :value="hour"
      @timeUpdate="updateHour"
      placeholder="hh"
    />

    <Time type="minute" :value="minute" @timeUpdate="updateMinute" placeholder="mm" />

    <Time
      v-if="shouldRenderSeconds"
      type="second"
      :value="second"
      @timeUpdate="updateSecond"
      placeholder="ss"
    />

    <Time
      v-if="showMeridian"
      type="meridian"
      :value="meridian"
      @timeUpdate="updateMeridian"
      placeholder="am"
    />
  </div>
</template>

<script>
import TimeRange from "./mixins/TimeRange";
import Time from "./Time/Time.vue";

export default {
  name: "FdTimeItem",
  mixins: [TimeRange],
  components: { Time },
  props: {
    ariaLabel: { type: String, default: "Time Item" },
    value: String, // validator needs to be added.
    showMeridian: { type: Boolean, default: false }
  },
  computed: {
    shouldRenderSeconds() {
      return this.second != null;
    }
  },
  watch: {
    timeValue: {
      immediate: true,
      handler(newTimeValue) {
        if (newTimeValue != null) {
          const time = newTimeValue.toString().split(" ");
          const timeValue = time[0].split(":");
          [this.hour, this.minute, this.second, this.meridian] = [
            timeValue[0],
            timeValue[1],
            timeValue[2],
            time[1]
          ];
        }
      }
    }
  },
  methods: {
    updateHour(hour) {
      const hourType = this.showMeridian ? "hour12" : "hour24";
      if (
        (Number(this.hour) === Number(this.range[hourType].max) &&
          Number(hour) === Number(this.range[hourType].max) - 1) ||
        (Number(hour) === Number(this.range[hourType].max) &&
          Number(this.hour) === Number(this.range[hourType].max) - 1)
      ) {
        let meridian = this.meridian;
        meridian =
          this.meridian === this.range.meridian.min
            ? this.range.meridian.max
            : this.range.meridian.min;
        this.updateMeridian(meridian);
      }
      this.validateTimeUnit("hour", hour, hourType);
    },
    updateMinute(minute) {
      const minuteValue = this.minute === "00" && minute === "-1" ? this.range.minute.max : minute;
      const updateParameters = {
        timeUnitType: "minute",
        value: minuteValue,
        timeType: "minute",
        upperTimeUnitType: "hour",
        upperTimeUnitUpdateHandler: this.updateHour
      };
      this.updateUnit(updateParameters);
    },
    updateSecond(second) {
      const updateParameters = {
        timeUnitType: "second",
        value: second,
        timeType: "second",
        upperTimeUnitType: "minute",
        upperTimeUnitUpdateHandler: this.updateMinute
      };
      this.updateUnit(updateParameters);
    },
    updateMeridian(meridian) {
      this.meridian = meridian ? meridian : "--";
      this.updateTimeItem();
    },
    updateUnit(updateParameters) {
      let dividend = 0;
      let remainder = 0;
      if (
        Number(this[updateParameters.timeUnitType]) ===
          Number(this.range[updateParameters.timeType].min) &&
        Number(updateParameters.value) === Number(this.range[updateParameters.timeType].max)
      ) {
        dividend = -1;
        remainder = Number(this.range[updateParameters.timeType].max);
        this[updateParameters.timeUnitType] = updateParameters.value;
      } else if (
        Number(this[updateParameters.timeUnitType]) ===
          Number(this.range[updateParameters.timeType].max) &&
        Number(updateParameters.value) === Number(this.range[updateParameters.timeType].min)
      ) {
        dividend = 1;
        remainder = Number(this.range[updateParameters.timeType].min);
        this[updateParameters.timeUnitType] = updateParameters.value;
      } else {
        if (
          this.validateTimeUnit(
            updateParameters.timeUnitType,
            updateParameters.value,
            updateParameters.timeType
          )
        ) {
          remainder =
            Number(updateParameters.value) %
            (Number(this.range[updateParameters.timeType].max) + 1);
          dividend = Math.floor(
            Number(updateParameters.value) / (Number(this.range[updateParameters.timeType].max) + 1)
          );
          this.validateTimeUnit(
            updateParameters.timeUnitType,
            remainder.toString().padStart(2, "0"),
            updateParameters.timeType
          );
        }
      }
      const upperValue = (Number(this[updateParameters.upperTimeUnitType]) + dividend)
        .toString()
        .padStart(2, "0");
      updateParameters.upperTimeUnitUpdateHandler(upperValue);
    },
    validateTimeUnit(unitType, value, type) {
      let valueOutOfRange = false;
      if (this.checkValueRange(value, type)) {
        this[unitType] = value ? value : "--";
      } else if (!isNaN(Number(value))) {
        valueOutOfRange = true;
      } else {
        this[unitType] = "--";
      }
      this.updateTimeItem();
      return valueOutOfRange;
    },
    updateTimeItem() {
      const timeValue = [this.hour, this.minute, this.second].join(":");
      const time = [timeValue, this.meridian].join(" ");
      this.timeValue = time;
      this.$emit("timeItemUpdate", this.timeValue);
    }
  },
  data() {
    return {
      timeValue: this.value,
      hour: "--",
      minute: "--",
      second: "--",
      meridian: "--"
    };
  }
};
</script>
