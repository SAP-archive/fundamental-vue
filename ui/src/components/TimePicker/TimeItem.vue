<template>
  <div class="fd-time">
    <Time
      :type="showMeridian ? 'hour12' : 'hour24'"
      :value="hour"
      @timeUpdate="updateHour"
      placeholder="hh"
    />

    <Time
      type="minute"
      :value="minute"
      @timeUpdate="updateMinute"
      placeholder="mm"
    />

    <Time
      v-if="shouldRenderSeconds"
      type="second"
      :value="second"
      @timeUpdate="updateSecond"
      placeholder="ss"
    />

    <Time
      v-if="showMeridian"
      type="meridian"
      :value="meridian"
      @timeUpdate="updateMeridian"
      placeholder="am"
    />
  </div>
</template>

<script lang="ts">
import { mixins } from "@/mixins";
import TimeRange from "./mixins/TimeRange";
import { TimeType } from "./Time/TimeType";
import Time from "./Time/Time.vue";

const timeUnitMapping = {
  hour: "hour",
  minute: "minute",
  second: "second",
  meridian: "meridian"
};

type TimeUnitType = keyof (typeof timeUnitMapping);
interface TimeUnitUpdateParams {
  timeUnitType: TimeUnitType;
  value: string;
  timeType: TimeType;
  upperTimeUnitType: TimeUnitType;
  upperTimeUnitUpdateHandler: (arg1: string) => void;
}

export default mixins(TimeRange).extend({
  name: "FdTimeItem",
  components: { Time },
  props: {
    ariaLabel: { type: String, default: "Time Item" },
    value: String, // validator needs to be added.
    showMeridian: { type: Boolean, default: false }
  },
  computed: {
    shouldRenderSeconds(): boolean {
      return this.second != null;
    }
  },
  watch: {
    timeValue: {
      immediate: true,
      handler(newTimeValue: string | null) {
        if (newTimeValue != null) {
          const time = newTimeValue.toString().split(" ");
          const timeValue = time[0].split(":");
          [this.hour, this.minute, this.second, this.meridian] = [
            timeValue[0],
            timeValue[1],
            timeValue[2],
            time[1]
          ];
        }
      }
    }
  },
  methods: {
    updateHour(hour: string) {
      const hourType = this.showMeridian ? "hour12" : "hour24";
      if (
        (Number(this.hour) === Number(this.range[hourType].max) &&
          Number(hour) === Number(this.range[hourType].max) - 1) ||
        (Number(hour) === Number(this.range[hourType].max) &&
          Number(this.hour) === Number(this.range[hourType].max) - 1)
      ) {
        let meridian = this.meridian;
        meridian =
          this.meridian === this.range.meridian.min
            ? this.range.meridian.max
            : this.range.meridian.min;
        this.updateMeridian(meridian);
      }
      this.validateTimeUnit("hour", hour, hourType);
    },
    updateMinute(minute: string) {
      const minuteValue =
        this.minute === "00" && minute === "-1"
          ? this.range.minute.max
          : minute;
      const updateParameters: TimeUnitUpdateParams = {
        timeUnitType: "minute",
        value: minuteValue,
        timeType: "minute",
        upperTimeUnitType: "hour",
        upperTimeUnitUpdateHandler: this.updateHour
      };
      this.updateUnit(updateParameters);
    },
    updateSecond(second: string) {
      const updateParameters: TimeUnitUpdateParams = {
        timeUnitType: "second",
        value: second,
        timeType: "second",
        upperTimeUnitType: "minute",
        upperTimeUnitUpdateHandler: this.updateMinute
      };
      this.updateUnit(updateParameters);
    },
    updateMeridian(meridian: string) {
      this.meridian = meridian ? meridian : "--";
      this.updateTimeItem();
    },
    updateUnit(updateParameters: TimeUnitUpdateParams) {
      let dividend = 0;
      let remainder = 0;
      if (
        Number(this[updateParameters.timeUnitType]) ===
          Number(this.range[updateParameters.timeType].min) &&
        Number(updateParameters.value) ===
          Number(this.range[updateParameters.timeType].max)
      ) {
        dividend = -1;
        remainder = Number(this.range[updateParameters.timeType].max);
        this[updateParameters.timeUnitType] = updateParameters.value;
      } else if (
        Number(this[updateParameters.timeUnitType]) ===
          Number(this.range[updateParameters.timeType].max) &&
        Number(updateParameters.value) ===
          Number(this.range[updateParameters.timeType].min)
      ) {
        dividend = 1;
        remainder = Number(this.range[updateParameters.timeType].min);
        this[updateParameters.timeUnitType] = updateParameters.value;
      } else {
        if (
          this.validateTimeUnit(
            updateParameters.timeUnitType,
            updateParameters.value,
            updateParameters.timeType
          )
        ) {
          remainder =
            Number(updateParameters.value) %
            (Number(this.range[updateParameters.timeType].max) + 1);
          dividend = Math.floor(
            Number(updateParameters.value) /
              (Number(this.range[updateParameters.timeType].max) + 1)
          );
          this.validateTimeUnit(
            updateParameters.timeUnitType,
            remainder.toString().padStart(2, "0"),
            updateParameters.timeType
          );
        }
      }
      const upperValue = (
        Number(this[updateParameters.upperTimeUnitType]) + dividend
      )
        .toString()
        .padStart(2, "0");
      updateParameters.upperTimeUnitUpdateHandler(upperValue);
    },
    validateTimeUnit(unitType: TimeUnitType, value: string, type: TimeType) {
      let valueOutOfRange = false;
      if (this.checkValueRange(value, type)) {
        this[unitType] = value ? value : "--";
      } else if (!isNaN(Number(value))) {
        valueOutOfRange = true;
      } else {
        this[unitType] = "--";
      }
      this.updateTimeItem();
      return valueOutOfRange;
    },
    updateTimeItem() {
      const timeValue = [this.hour, this.minute, this.second].join(":");
      const time = [timeValue, this.meridian].join(" ");
      this.timeValue = time;
      this.$emit("timeItemUpdate", this.timeValue);
    }
  },
  data() {
    return {
      timeValue: this.value as string | number | null,
      hour: "--" as string | number,
      minute: "--" as string | number,
      second: "--" as string | number,
      meridian: "--" as string | number
    };
  }
});
</script>
