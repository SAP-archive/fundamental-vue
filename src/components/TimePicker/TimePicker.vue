<template>
  <div class="fd-time-picker">
    <FdPopover adjustsBodyWidth placement="bottom-start">
      <template #control="{toggle, show, hide}">
        <InputGroup
          :compact="compact"
          afterClass="fd-input-group__addon--button"
        >
          <Input
            :value="timeValue"
            :compact="compact"
            :placeholder="placeholder"
          />
          <Button
            @click="toggle"
            styling="light"
            slot="after"
            icon="fob-watch"
          />
        </InputGroup>
      </template>
      <template #default>
        <TimeItem
          :value="timeValue"
          :showMeridian="showMeridian"
          @timeItemUpdate="updateTimeItem"
        />
      </template>
    </FdPopover>
  </div>
</template>

<script>
import FdPopover from "./../Popover/Popover.vue";
import { InputGroup } from "./../Form";
import Input from "./../Form/Controls/Input.vue";
import { Button } from "./../Button";
import TimeItem from "./TimeItem.vue";

export default {
  name: "FdTimePicker",
  components: { FdPopover, Input, InputGroup, Button, TimeItem },
  props: {
    id: String, // FIXME
    placeholder: String,
    value: String, // FIXME: Add Validator
    ariaLabel: String,
    compact: { type: Boolean, default: false },
    showMeridian: { type: Boolean, default: false }
  },
  methods: {
    updateTimeItem(time) {
      this.timeValue = time;
      this.$emit("timePickerUpdate", this.timeValue);
    }
  },
  data() {
    return {
      timeValue: this.value
    };
  }
};
</script>
