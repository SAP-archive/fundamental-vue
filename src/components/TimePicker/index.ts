import TimePicker from "./TimePicker.vue";
import TimeItem from "./TimeItem.vue";
import TimeInput from "./Time/TimeInput.vue";
import TimeAction from "./Time/TimeAction.vue";
import Time from "./Time/Time.vue";
import { pluginify } from "@/util";
export default pluginify(TimePicker, TimeItem, TimeInput, TimeAction, Time);
export { TimePicker, TimeItem, TimeInput, TimeAction, Time };
