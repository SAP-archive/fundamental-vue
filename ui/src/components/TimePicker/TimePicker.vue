<template>
  <div class="fd-time-picker">
    <Popover :uid="uid" noArrow :popoverVisible="false">
      <div class="fd-popover__control" slot="control">
        <InputGroup
          :compact="compact"
          afterClass="fd-input-group__addon--button fd-input-group--after"
        >
          <Input
            :value="timeValue"
            :compact="compact"
            :placeholder="placeholder"
          />
          <Button styling="light" slot="after" icon="fob-watch" />
        </InputGroup>
      </div>
      <TimeItem
        :value="timeValue"
        :showMeridian="showMeridian"
        @timeItemUpdate="updateTimeItem"
      />
    </Popover>
  </div>
</template>

<script lang="ts">
import { Uid, mixins } from "@/mixins";
import { Popover } from "@/components/Popover";
import { Input, InputGroup } from "@/components/Form";
import { Button } from "@/components/Button";
import TimeItem from "./TimeItem.vue";

export default mixins(Uid).extend({
  name: "FdTimePicker",
  components: { Popover, Input, InputGroup, Button, TimeItem },
  props: {
    id: String, // FIXME
    placeholder: String,
    value: String, // FIXME: Add Validator
    ariaLabel: String,
    compact: { type: Boolean, default: false },
    showMeridian: { type: Boolean, default: false }
  },
  methods: {
    updateTimeItem(time: string): void {
      this.timeValue = time;
      this.$emit("timePickerUpdate", this.timeValue);
    }
  },
  data() {
    return {
      timeValue: this.value as string | null
    };
  }
});
</script>
