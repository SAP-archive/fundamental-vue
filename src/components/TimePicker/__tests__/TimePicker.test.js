import { mount } from "@vue/test-utils";
import TimePicker from "../TimePicker.vue";

jest.mock("./../../../lib/uuid", () => {
  return {
    shortUuid() {
      return "time-picker";
    }
  };
});

describe("TimePicker", () => {
  const placeholder = "Pick a time";
  const value = "11:01:58 am";
  const ariaLabel = "Clock In Time";

  it("renders correctly", async () => {
    const wrapper = mount(TimePicker, {
      propsData: {
        uid: "time-picker",
        placeholder,
        value,
        ariaLabel
      }
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });

  it("shows up the picker", async () => {
    const wrapper = mount(TimePicker, {
      propsData: {
        uid: "time-picker",
        placeholder,
        value,
        ariaLabel
      }
    });
    await wrapper.vm.$nextTick();
    const popoverBody = wrapper.find(".fd-popover__popper");
    expect(popoverBody.attributes("aria-hidden")).toBe("true");

    const inputGroup = wrapper.find(".fd-input-group");
    inputGroup.trigger("click");

    expect(popoverBody.attributes("aria-hidden")).toBe("true");
  });

  describe("Compact Style", async () => {
    it("renders correctly", async () => {
      const wrapper = mount(TimePicker, {
        propsData: {
          uid: "time-picker",
          compact: true,
          showMeridian: false
        }
      });
      await wrapper.vm.$nextTick();
      expect(wrapper.element).toMatchSnapshot();
    });

    it("has compact style", async () => {
      const wrapper = mount(TimePicker, {
        propsData: {
          uid: "time-picker",
          compact: true,
          showMeridian: false
        }
      });
      await wrapper.vm.$nextTick();
      expect(wrapper.contains(".fd-input--compact")).toBe(true);
    });

    it("does not have meridian", async () => {
      const wrapper = mount(TimePicker, {
        propsData: {
          uid: "time-picker",
          compact: true,
          showMeridian: false
        }
      });
      await wrapper.vm.$nextTick();
      expect(wrapper.findAll(".fd-time__input")).toHaveLength(3);
    });
  });

  describe("With Meridian", () => {
    it("renders correctly", async () => {
      const wrapper = mount(TimePicker, {
        propsData: {
          uid: "time-picker",
          showMeridian: true,
          compact: false
        }
      });
      await wrapper.vm.$nextTick();
      expect(wrapper.element).toMatchSnapshot();
    });

    it("has meridian", async () => {
      const wrapper = mount(TimePicker, {
        propsData: {
          uid: "time-picker",
          showMeridian: true,
          compact: false
        }
      });
      await wrapper.vm.$nextTick();
      expect(wrapper.findAll(".fd-time__input")).toHaveLength(4);
    });

    it("does not have compact style", async () => {
      const wrapper = mount(TimePicker, {
        propsData: {
          uid: "time-picker",
          showMeridian: true,
          compact: false
        }
      });
      await wrapper.vm.$nextTick();
      expect(wrapper.contains(".fd-input--compact")).toBe(false);
    });
  });
});
