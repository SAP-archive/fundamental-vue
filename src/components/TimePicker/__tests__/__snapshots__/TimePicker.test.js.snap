// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker Compact Style renders correctly 1`] = `
<div
  class="fd-time-picker"
  uid="time-picker"
>
  <div
    class="fd-popover"
  >
    <div
      class="fd-popover__control"
    >
      <div
        class="fd-input-group fd-input-group--after fd-input-group--compact"
      >
        <!---->
         
        <input
          class="fd-input fd-form__control fd-input--compact"
          placeholder=""
          type="text"
        />
          
        <span
          class="fd-input-group__addon fd-input-group__addon--after fd-input-group--compact fd-input-group__addon--button"
        >
          <button
            class="fd-button--light sap-icon--fob-watch"
          />
        </span>
      </div>
    </div>
     
  </div>
</div>
`;

exports[`TimePicker With Meridian renders correctly 1`] = `
<div
  class="fd-time-picker"
  uid="time-picker"
>
  <div
    class="fd-popover"
  >
    <div
      class="fd-popover__control"
    >
      <div
        class="fd-input-group fd-input-group--after"
      >
        <!---->
         
        <input
          class="fd-input fd-form__control"
          placeholder=""
          type="text"
        />
          
        <span
          class="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button"
        >
          <button
            class="fd-button--light sap-icon--fob-watch"
          />
        </span>
      </div>
    </div>
     
  </div>
</div>
`;

exports[`TimePicker renders correctly 1`] = `
<div
  class="fd-time-picker"
  uid="time-picker"
>
  <div
    class="fd-popover"
  >
    <div
      class="fd-popover__control"
    >
      <div
        class="fd-input-group fd-input-group--after"
      >
        <!---->
         
        <input
          class="fd-input fd-form__control"
          placeholder="Pick a time"
          type="text"
        />
          
        <span
          class="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button"
        >
          <button
            class="fd-button--light sap-icon--fob-watch"
          />
        </span>
      </div>
    </div>
     
  </div>
</div>
`;
